### PR TITLE
Add ceph support

### PR DESCRIPTION
**Pull request by tolkonepiu**
Add support for [Ceph](http://ceph.com/)

### CODE REVIEW

File: modules/ceph/pom.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>testcontainers</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-s3</artifactId>
```
Comment (bsideup) [inline modules/ceph/pom.xml]: please declare it as `provided` dependency

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        withEnv("RGW_NAME", rgwName);
+        withEnv("NETWORK_AUTO_DETECT", networkAutoDetectMode.value);
+        withEnv("CEPH_DEMO_UID", demoUid);
+        withEnv("CEPH_DEMO_ACCESS_KEY", awsAccessKey);
+        withEnv("CEPH_DEMO_SECRET_KEY", awsSecretKey);
+        withEnv("CEPH_DEMO_BUCKET", bucketName);
+        withExposedPorts(6789, 6800, 6801, 6802, 6803, 6804, 6805, 80, 5000);
+        waitingFor(
+                new LogMessageWaitStrategy()
+                        .withRegEx(".*\\/entrypoint.sh: SUCCESS\n")
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: Ceph has a built-in health indicator, please use Http wait strategy with `/api/v0.1/health` on port 5000

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    private String bucketName = "CEPH";
+
+    private String rgwName = "localhost";
+
+    private String demoUid = "ceph";
+
+    private NetworkAutoDetectMode networkAutoDetectMode = NetworkAutoDetectMode.IPV4_ONLY;
+
+    private Duration startupTimeout = Duration.ofSeconds(20);
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: TestContainers already declares startup timeout inside `GenericContainer`

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    public CephContainer() {
+        super(IMAGE + ":latest");
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: please don't use `:latest`, choose some concrete tag instead: https://hub.docker.com/r/ceph/demo/tags/

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    public enum NetworkAutoDetectMode {
+        IPV6_OR_IPV4("1"),
+        IPV4_ONLY("4"),
+        IPV6_ONLY("6");
+
+        private String value;
+
+        NetworkAutoDetectMode(String value) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: can be replaced with `@RequiredArgumentsConstructor` if `value` is `final` (which is also good anyway)

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        return bucketName;
+    }
+
+    public SELF withBucketName(String bucketName) {
+        //because s3cmd transforming bucket name to uppercase
+        this.bucketName = bucketName.toUpperCase();
+        return self();
+    }
+
+    public String getRgwName() {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: please replace getters with Lombok's `@Getter` annotation

File: modules/ceph/src/test/java/org/testcontainers/containers/CephContainerTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                .withClientConfiguration(
+                        new ClientConfiguration()
+                                .withProtocol(Protocol.HTTP)
+                                .withSignerOverride("S3SignerType")
+                )
+                .build();
+    }
+
+    @Test
+    public void testS3Bucket() {
```
Comment (bsideup) [inline modules/ceph/src/test/java/org/testcontainers/containers/CephContainerTest.java]: I think this test does too much, so the Ceph integration we only need to check that API works in general :)

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 2 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: it's not obvious
```java
protected Integer getLivenessCheckPort() {
        if (exposedPorts.size() > 0) {
            return getMappedPort(exposedPorts.get(0));
        } else if (portBindings.size() > 0) {
            return Integer.valueOf(PortBinding.parse(portBindings.get(0)).getBinding().getHostPortSpec());
        } else {
            return null;
        }
    }
```

File: modules/ceph/src/test/java/org/testcontainers/containers/CephContainerTest.java (hunk 7 above)
Comment (tolkonepiu) [reply modules/ceph/src/test/java/org/testcontainers/containers/CephContainerTest.java]: I added these tests for an example of connecting s3-sdk to ceph.

**Comment by tolkonepiu** [comment]
@rnorth done

**Comment by tolkonepiu** [comment]
@rnorth done

**Comment by rnorth** [comment]
We should keep this open. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by kiview** [comment]
@rnorth Something missing for you? Seems good to merge.

**Comment by bsideup** [comment]
@kiview we should remove at least the self typing here

File: modules/ceph/build.gradle [hunk 8]
```diff
@@ -0,0 +1,7 @@
+description = "Testcontainers :: JDBC :: Ceph"
```
Comment (bsideup) [inline modules/ceph/build.gradle]:
```suggestion
description = "Testcontainers :: Ceph"
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.testcontainers.containers.wait.HttpWaitStrategy;
+
+import java.time.Duration;
+
+@Getter
+public class CephContainer<SELF extends CephContainer<SELF>> extends GenericContainer<SELF> {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
public class CephContainer extends GenericContainer< CephContainer > {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public AwsClientBuilder.EndpointConfiguration getAWSEndpointConfiguration() {
+        return new AwsClientBuilder.EndpointConfiguration(
+                getContainerIpAddress() + ":" + getMappedPort(S3_PORT),
+                "us-east-1"
+        );
+    }
+
+    public SELF withAwsAccessKey(String awsAccessKey) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withAwsAccessKey(String awsAccessKey) {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        this.awsAccessKey = awsAccessKey;
+        return self();
+    }
+
+    public SELF withAwsSecretKey(String awsSecretKey) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withAwsSecretKey(String awsSecretKey) {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        this.awsSecretKey = awsSecretKey;
+        return self();
+    }
+
+    public SELF withBucketName(String bucketName) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withBucketName(String bucketName) {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        //because s3cmd transforming bucket name to uppercase
+        this.bucketName = bucketName.toUpperCase();
+        return self();
+    }
+
+    public SELF withRgwName(String rgwName) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withRgwName(String rgwName) {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        this.rgwName = rgwName;
+        return self();
+    }
+
+    public SELF withDemoUid(String demoUid) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withDemoUid(String demoUid) {
```

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        this.demoUid = demoUid;
+        return self();
+    }
+
+    public SELF withNetworkAutoDetectMode(NetworkAutoDetectMode networkAutoDetectMode) {
```
Comment (bsideup) [inline modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]:
```suggestion
    public CephContainer withNetworkAutoDetectMode(NetworkAutoDetectMode networkAutoDetectMode) {
```

Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 14 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 13 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 12 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 11 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

File: modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java (hunk 10 above)
Comment (tolkonepiu) [reply modules/ceph/src/main/java/org/testcontainers/containers/CephContainer.java]: 736437e4641c85385fbdf658391518cf9c4af810

**Comment by tolkonepiu** [comment]
@bsideup ping

**Comment by rnorth** [comment]
FWIW I've raised this ticket around the current gap in our docs for new module contributors. I'd like to have a bit of a think about the third point specifically before merging this, but otherwise I'm fine with this new module at an overall level: https://github.com/testcontainers/testcontainers-java/issues/1503

**Comment by rnorth** [comment]
Not stale - reopening.

**Comment by rnorth** [comment]
I've made some tweaks but cannot push to the original branch, so have raised #2687. Will close this original PR.

**Pull request closed without merging**